microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/model/PropertyGraph.java

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+* Copyright (c) 2020 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.lsp4mp.model;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Queue;
+
+public class PropertyGraph {
+
+	private Map<String /* name of property */, List<String> /* adjacency list */> nodes;
+
+	public PropertyGraph() {
+		this.nodes = new HashMap<String, List<String>>();
+	}
+
+	public void addNode(String property) {
+		this.nodes.put(property, new ArrayList<String>());
+	}
+
+	public void addEdge(String from, String to) {
+		if (hasNode(from) && hasNode(to)) {
+			this.nodes.get(from).add(to);
+		}
+	}
+
+	public boolean hasNode(String property) {
+		return this.nodes.containsKey(property);
+	}
+
+	/**
+	 * Gets a list of properties that do not depend on <code>property</code>.
+	 * 
+	 * These are the properties that, if a dependency of <code>property</code> on
+	 * them were introduced, wouldn't introduce a circular dependency.
+	 * 
+	 * @param property The property the find the independent properties of.
+	 * @return A list of all the properties whose value do not depend on this
+	 *         property.
+	 */
+	public List<String> independentProperties(String property) {
```

Comment:
I wasn't sure if this method belongs here because I was thinking `PropertyGraph` is a simple Graph class that manages edges and could be reused for different applications (but maybe it's not very likely that we need graphs again).

But on second thought, maybe I'm wrong, since `PropertyGraph` was created specifically to determine relationships between properties and to solve problems like this "independent properties" problem.

Reply:
Hmm. I don't know either way as well. It would be easy enough to move `independentProperties` to a utilities class if it makes sense to do this.